lib/crystalball/graphql/map_generator/graphql_strategy.rb

```diff
@@ -0,0 +1,125 @@
+# frozen_string_literal: true
+
+require 'crystalball/map_generator/base_strategy'
+
+module Crystalball
+  module GraphQL
+    module MapGenerator
+      # Map generator strategy to build map of GraphQL type definitions
+      # that might affect an example.
+      # Patches `GraphQL::ObjectType.define` to get the path of files that call it.
+      class GraphQLStrategy
+        include ::Crystalball::MapGenerator::BaseStrategy
+        include ::Crystalball::MapGenerator::Helpers::PathFilter
+
+        class << self
+          attr_accessor :current_schema, :type_definition_paths, :patch_applied
+
+          # Adds the type definition paths of each of the `other_keys`
+          # to the type definition paths of `key`
+          def augment_type_definition_paths(key, other_keys)
+            type_definition_paths[key.object_id] |= other_keys
+                                                    .map { |t| type_definition_paths[t.object_id] }
+                                                    .flatten
+                                                    .compact
+          end
+
+          # Checks whether `::GraphQL::Schema` is defined and, if it is, applies
+          # the patch to it. It needs to be this way because `graphql` gem is weird.
+          def apply_patch
+            return unless defined?(::GraphQL::Schema)
+
+            ::GraphQL::Schema.class_eval do
+              prepend Patch
+            end
+
+            self.patch_applied = true
+          end
+
+          # Same as `apply_patch` but raises error if `::GraphQL::Schema` is not defined
+          # @see Crystalball::GraphQL::MapGenerator::GraphQLStrategy.apply_patch
+          def apply_patch!
+            apply_patch || raise(
+              NameError,
+              "::GraphQL::Schema is not defined. You need to `require 'graphql'` in order to use this strategy"
+            )
+          end
+
+          def type_definition_paths_during
+            apply_patch! unless patch_applied
+            self.current_schema = nil
+
+            yield
+
+            return [] unless current_schema
+            type_definition_paths[current_schema.object_id]
+          end
+        end
+
+        attr_accessor :root
+
+        # Tracer to run on every GraphQL query.
```

Comment:
will it work for cases where we don't execute a query? E.g. when we just load the type of a schema

Reply:
No, but there's no point, or at least I haven't thought of any situations where you can use the type in any other way. If the example does not execute any queries, you're probably testing graphql types in a weird way and it will be caught by associated_specs strategy when executing.